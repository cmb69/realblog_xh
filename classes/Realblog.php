<?php

/**
 * Copyright 2006-2010 Jan Kanters
 * Copyright 2010-2014 Gert Ebersbach
 * Copyright 2014-2017 Christoph M. Becker
 *
 * This file is part of Realblog_XH.
 *
 * Realblog_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Realblog_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Realblog_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Realblog;

use stdClass;
use ReflectionClass;

class Realblog
{
    const VERSION = '3.0beta9';

    /**
     * @return void
     */
    public static function init()
    {
        global $su, $plugin_cf, $plugin_tx;

        self::registerCommands();
        if ($plugin_cf['realblog']['auto_publish']) {
            self::autoPublish();
        }
        if ($plugin_cf['realblog']['auto_archive']) {
            self::autoArchive();
        }
        if ($plugin_cf['realblog']['rss_enabled']) {
            self::emitAlternateRSSLink();
            $rssFeedRequested = filter_input(
                INPUT_GET,
                'realblog_feed',
                FILTER_VALIDATE_REGEXP,
                array('options' => array('regexp' => '/^rss$/'))
            );
            if ($rssFeedRequested) {
                $controller = new FeedController($plugin_cf['realblog'], $plugin_tx['realblog'], new Finder());
                echo $controller->defaultAction();
                exit;
            }
        }
        /** @psalm-suppress UndefinedConstant */
        if (XH_ADM) {
            self::registerPluginMenu();
            if (XH_wantsPluginAdministration('realblog') || $su === 'realblog') {
                self::handleAdministration();
            }
        }
    }

    /**
     * @return void
     */
    private static function registerCommands()
    {
        $class = new ReflectionClass(Realblog::class);
        $commands = array();
        foreach ($class->getMethods() as $method) {
            $methodName = $method->getName();
            if (preg_match('/.*(?=Command$)/', $methodName, $m)) {
                $commands[$m[0]] = $method->getParameters();
            }
        }
        foreach ($commands as $name => $params) {
            $paramList = $argList = array();
            foreach ($params as $param) {
                $string = '$' . $param->getName();
                $argList[] = $string;
                if ($param->isOptional()) {
                    $string .= '=' . var_export($param->getDefaultValue(), true);
                }
                $paramList[] = $string;
            }
            $paramString = implode(',', $paramList);
            $argString = implode(',', $argList);
            eval("function Realblog_$name($paramString) {return \\Realblog\\Realblog::{$name}Command($argString);}");
        }
    }

    /**
     * @return void
     */
    private static function registerPluginMenu()
    {
        global $sn, $plugin_tx;

        XH_registerStandardPluginMenuItems(true);
        XH_registerPluginMenuItem(
            'realblog',
            $plugin_tx['realblog']['exchange_heading'],
            "$sn?realblog&admin=data_exchange"
        );
    }

    /**
     * @return void
     */
    private static function handleAdministration()
    {
        global $sn, $admin, $o, $plugin_tx;

        $o .= print_plugin_admin('on');
        pluginMenu('ROW');
        pluginMenu('TAB', "$sn?realblog&admin=data_exchange", '', $plugin_tx['realblog']['exchange_heading']);
        $o .= pluginMenu('SHOW');
        switch ($admin) {
            case '':
                $o .= (new InfoController(new View()))->defaultAction();
                break;
            case 'plugin_main':
                self::routeTo(MainAdminController::class);
                break;
            case 'data_exchange':
                self::routeTo(DataExchangeController::class);
                break;
            default:
                $o .= plugin_admin_common();
        }
    }

    /**
     * @param class-string $controllerClassName
     * @return void
     */
    private static function routeTo($controllerClassName)
    {
        global $o, $action, $plugin_cf, $plugin_tx, $_XH_csrfProtection;

        $methodName = lcfirst(implode('', array_map('ucfirst', explode('_', $action)))) . 'Action';
        $controller = new $controllerClassName(
            $plugin_cf['realblog'],
            $plugin_tx['realblog'],
            new Finder(),
            $_XH_csrfProtection,
            new View()
        );
        $class = new ReflectionClass($controller);
        if ($class->hasMethod($methodName)
            && ($method = $class->getMethod($methodName))
            && $method->isPublic()
        ) {
            $o .= $method->invoke($controller);
        } else {
            $o .= $controller->defaultAction();
        }
    }

    /**
     * @return void
     * @global string $hjs
     */
    private static function emitAlternateRSSLink()
    {
        global $hjs;

        $hjs .= '<link rel="alternate" type="application/rss+xml"'
            . ' href="./?realblog_feed=rss">';
    }

    /**
     * @return void
     */
    private static function autoPublish()
    {
        DB::autoChangeStatus('publishing_date', 1);
    }

    /**
     * @return void
     */
    private static function autoArchive()
    {
        DB::autoChangeStatus('archiving_date', 2);
    }

    /**
     * @param bool $showSearch
     * @param string $category
     * @return string
     */
    public static function blogCommand($showSearch = false, $category = 'all')
    {
        global $plugin_cf, $plugin_tx;

        $controller = new BlogController(
            $plugin_cf['realblog'],
            $plugin_tx['realblog'],
            $showSearch,
            new Finder(),
            new View(),
            $category
        );
        if (filter_has_var(INPUT_GET, 'realblog_id')) {
            return (string) $controller->showArticleAction(filter_input(
                INPUT_GET,
                'realblog_id',
                FILTER_VALIDATE_INT,
                array('options' => array('min_range' => 1))
            ));
        } else {
            return $controller->defaultAction();
        }
    }

    /**
     * @param bool $showSearch
     * @return string
     */
    public static function archiveCommand($showSearch = false)
    {
        global $plugin_cf, $plugin_tx;

        $controller = new ArchiveController(
            $plugin_cf['realblog'],
            $plugin_tx['realblog'],
            $showSearch,
            new Finder(),
            new View()
        );
        if (filter_has_var(INPUT_GET, 'realblog_id')) {
            return (string) $controller->showArticleAction(filter_input(
                INPUT_GET,
                'realblog_id',
                FILTER_VALIDATE_INT,
                array('options' => array('min_range' => 1))
            ));
        } else {
            return $controller->defaultAction();
        }
    }

    /**
     * @param string $pageUrl
     * @param bool $showTeaser
     * @return string
     */
    public static function linkCommand($pageUrl, $showTeaser = false)
    {
        global $plugin_cf, $plugin_tx;

        $controller = new LinkController(
            $plugin_cf['realblog'],
            $plugin_tx['realblog'],
            $pageUrl,
            $showTeaser,
            new Finder(),
            new View()
        );
        return $controller->defaultAction();
    }

    /**
     * @param string $pageUrl
     * @return string
     */
    public static function mostPopularCommand($pageUrl)
    {
        global $plugin_cf, $plugin_tx;

        $controller = new MostPopularController(
            $plugin_cf['realblog'],
            $plugin_tx['realblog'],
            $pageUrl,
            new Finder(),
            new View()
        );
        return $controller->defaultAction();
    }

    /**
     * @param string $target
     * @return string
     */
    public static function feedLinkCommand($target = '_self')
    {
        global $plugin_cf, $plugin_tx;

        $controller = new FeedLinkController($plugin_cf['realblog'], $plugin_tx['realblog']);
        return $controller->defaultAction($target);
    }

    /**
     * @return int
     */
    public static function getPage()
    {
        global $edit;

        /** @psalm-suppress UndefinedConstant */
        if (XH_ADM && $edit) {
            if (filter_has_var(INPUT_GET, 'realblog_page')) {
                $page = filter_input(
                    INPUT_GET,
                    'realblog_page',
                    FILTER_VALIDATE_INT,
                    array('options' => array('min_range' => 1))
                );
                $_COOKIE['realblog_page'] = $page;
                setcookie('realblog_page', $page, 0, CMSIMPLE_ROOT);
            } else {
                $page = filter_input(
                    INPUT_COOKIE,
                    'realblog_page',
                    FILTER_VALIDATE_INT,
                    array('options' => array('min_range' => 1, 'default' => 1))
                );
            }
        } else {
            $page = filter_input(
                INPUT_GET,
                'realblog_page',
                FILTER_VALIDATE_INT,
                array('options' => array('min_range' => 1, 'default' => 1))
            );
        }
        return $page;
    }

    /**
     * @param int $num
     * @return bool
     */
    public static function getFilter($num)
    {
        $varname = "realblog_filter$num";
        if (filter_has_var(INPUT_GET, $varname)) {
            $filter = filter_input(INPUT_GET, $varname, FILTER_VALIDATE_BOOLEAN);
            $_COOKIE[$varname] = $filter ? 'on' : '';
            setcookie($varname, $filter ? 'on' : '', 0, CMSIMPLE_ROOT);
        } else {
            $filter = filter_input(INPUT_COOKIE, $varname, FILTER_VALIDATE_BOOLEAN);
        }
        return $filter;
    }

    /**
     * @param string $pageUrl
     * @param array  $params
     * @return string
     */
    public static function url($pageUrl, $params = array())
    {
        global $sn;

        $replacePairs = array(
            //'realblog_id' => 'id',
            //'realblog_page' => 'page'
        );
        $url = $sn . '?' . $pageUrl;
        ksort($params);
        foreach ($params as $name => $value) {
            if (!($name == 'realblog_page' && $value == 1
                || $name == 'realblog_year' && $value == date('Y')
                || $name == 'realblog_search' && !$value)
            ) {
                $url .= '&' . strtr($name, $replacePairs) . '=' . $value;
            }
        }
        return $url;
    }
}
