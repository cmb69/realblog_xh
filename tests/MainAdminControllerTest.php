<?php

/**
 * Copyright 2023 Christoph M. Becker
 *
 * This file is part of Realblog_XH.
 *
 * Realblog_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Realblog_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Realblog_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Realblog;

use ApprovalTests\Approvals;
use PHPUnit\Framework\TestCase;
use Realblog\Infra\DB;
use Realblog\Infra\FakeCsrfProtector;
use Realblog\Infra\FakeEditor;
use Realblog\Infra\FakeRequest;
use Realblog\Infra\Finder;
use Realblog\Infra\View;
use Realblog\Value\Article;
use Realblog\Value\FullArticle;

class MainAdminControllerTest extends TestCase
{
    public function testSetsPageCookie()
    {
        $sut = $this->sut();
        $request = new FakeRequest([
            "admin" => true,
            "edit" => true,
            "get" => ["realblog_page" => "3"],
            "path" => ["folder" => ["plugins" => "./plugins/"]],
        ]);
        $response = $sut($request, "");
        $this->assertEquals(["realblog_page" => "3"], $response->cookies());
    }

    public function testDefaultActionRendersOverview(): void
    {
        $sut = $this->sut(["finder" => ["articles" => $this->articles()]]);
        $request = new FakeRequest(["path" => ["folder" => ["plugins" => "./plugins/"]]]);
        $response = $sut($request, "");
        Approvals::verifyHtml($response->output());
    }

    public function testDefaultActionSetsFilterCookie()
    {
        $sut = $this->sut();
        $request = new FakeRequest([
            "path" => ["folder" => ["plugins" => "./plugins/"]],
            "get" => ["realblog_filter" => ["on", "", "on"]],
        ]);
        $response = $sut($request, "");
        $this->assertEquals(["realblog_filter" => "[true,false,true]"], $response->cookies());
    }

    public function testDefaultActionGetsPageAndFilterFromCookie()
    {
        $sut = $this->sut();
        $request = new FakeRequest([
            "path" => ["folder" => ["plugins" => "./plugins/"]],
            "admin" => true,
            "edit" => true,
            "cookie" => [
                "realblog_page" => "1",
                "realblog_filter" => "[true,false,true]",
            ],
        ]);
        $response = $sut($request, "");
        Approvals::verifyHtml($response->output());
    }

    public function testCreateActionRendersArticle(): void
    {
        $sut = $this->sut(["finder" => ["article" => $this->firstArticle()]]);
        $request = new FakeRequest([
            "path" => ["folder" => ["plugins" => "./plugins/"]],
            "server" => ["REQUEST_TIME" => 1675205155]
        ]);
        $response = $sut($request, "create");
        Approvals::verifyHtml($response->output());
        $this->assertEquals("Create new article", $response->title());
    }

    public function testCreateActionInitializesEditor(): void
    {
        $editor = new FakeEditor;
        $sut = $this->sut(["finder" => ["article" => $this->firstArticle()], "editor" => $editor]);
        $request = new FakeRequest([
            "path" => ["folder" => ["plugins" => "./plugins/"]],
            "server" => ["REQUEST_TIME" => 1675205155]
        ]);
        $sut($request, "create");
        $this->assertEquals(["realblog_headline_field", "realblog_story_field"], $editor->classes());
    }

    public function testCreateActionOutputsBjs(): void
    {
        $sut = $this->sut(["finder" => ["article" => $this->firstArticle()]]);
        $request = new FakeRequest([
            "path" => ["folder" => ["plugins" => "./plugins/"]],
            "server" => ["REQUEST_TIME" => 1675205155]
        ]);
        $response = $sut($request, "create");
        Approvals::verifyHtml($response->bjs());
    }

    public function testEditActionRendersArticle(): void
    {
        $sut = $this->sut(["finder" => ["article" => $this->firstArticle()]]);
        $request = new FakeRequest(["path" => ["folder" => ["plugins" => "./plugins/"]]]);
        $response = $sut($request, "edit");
        Approvals::verifyHtml($response->output());
        $this->assertEquals("Edit article #1", $response->title());
    }

    public function testEditActionInitializesEditor(): void
    {
        $editor = new FakeEditor;
        $sut = $this->sut(["finder" => ["article" => $this->firstArticle()], "editor" => $editor]);
        $request = new FakeRequest(["path" => ["folder" => ["plugins" => "./plugins/"]]]);
        $sut($request, "edit");
        $this->assertEquals(["realblog_headline_field", "realblog_story_field"], $editor->classes());
    }

    public function testEditActionOutputsBjs(): void
    {
        $sut = $this->sut(["finder" => ["article" => $this->firstArticle()]]);
        $request = new FakeRequest(["path" => ["folder" => ["plugins" => "./plugins/"]]]);
        $response = $sut($request, "edit");
        Approvals::verifyHtml($response->bjs());
    }

    public function testEditActionRendersArticleWithAutoInputs(): void
    {
        $sut = $this->sut([
            "conf" => ["auto_publish" => "true", "auto_archive" => "true"],
            "finder" => ["article" => $this->firstArticle()]
        ]);
        $request = new FakeRequest(["path" => ["folder" => ["plugins" => "./plugins/"]]]);
        $response = $sut($request, "edit");
        Approvals::verifyHtml($response->output());
    }

    public function testEditActionFailsOnMissingArticle(): void
    {
        $sut = $this->sut();
        $request = new FakeRequest();
        $response = $sut($request, "edit");
        Approvals::verifyHtml($response->output());
    }

    public function testDeleteActionRendersArticle(): void
    {
        $sut = $this->sut(["finder" => ["article" => $this->firstArticle()]]);
        $request = new FakeRequest(["path" => ["folder" => ["plugins" => "./plugins/"]]]);
        $response = $sut($request, "delete");
        Approvals::verifyHtml($response->output());
        $this->assertEquals("Delete article #1", $response->title());
    }

    public function testDeleteActionInitializesEditor(): void
    {
        $editor = new FakeEditor;
        $sut = $this->sut(["finder" => ["article" => $this->firstArticle()], "editor" => $editor]);
        $request = new FakeRequest(["path" => ["folder" => ["plugins" => "./plugins/"]]]);
        $sut($request, "delete");
        $this->assertEquals(["realblog_headline_field", "realblog_story_field"], $editor->classes());
    }

    public function testDeletectionOutputsBjs(): void
    {
        $sut = $this->sut(["finder" => ["article" => $this->firstArticle()]]);
        $request = new FakeRequest(["path" => ["folder" => ["plugins" => "./plugins/"]]]);
        $response = $sut($request, "delete");
        Approvals::verifyHtml($response->bjs());
    }

    public function testDeleteActionFailsOnMissingArticle(): void
    {
        $sut = $this->sut();
        $request = new FakeRequest();
        $response = $sut($request, "delete");
        Approvals::verifyHtml($response->output());
    }

    public function testDoCreateActionIsCsrfProtected()
    {
        $_POST = $this->dummyPost();
        $csrfProtector = new FakeCsrfProtector;
        $sut = $this->sut(["csrfProtector" => $csrfProtector, "db" => ["insert" => 0]]);
        $request = new FakeRequest(["server" => ["REQUEST_TIME" => 1675205155]]);
        $sut($request, "do_create");
        $this->assertTrue($csrfProtector->hasChecked());
    }

    public function testDoCreateActionRedirectsOnSuccess()
    {
        $_POST = $this->dummyPost();
        $sut = $this->sut(["db" => ["insert" => 1]]);
        $request = new FakeRequest(["server" => ["REQUEST_TIME" => 1675205155]]);
        $response = $sut($request, "do_create");
        $this->assertEquals(
            "http://example.com/?realblog&admin=plugin_main&action=plugin_text",
            $response->location()
        );
    }

    public function testDoCreateActionFailureIsReported(): void
    {
        $_POST = $this->dummyPost();
        $sut = $this->sut(["finder" => ["article" => $this->firstArticle()], "db" => ["insert" => 0]]);
        $request = new FakeRequest(["server" => ["REQUEST_TIME" => 1675205155]]);
        $response = $sut($request, "do_create");
        $this->assertEquals("Create new article", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testDoEditActionIsCsrfProtected()
    {
        $_POST = $this->dummyPost();
        $csrfProtector = new FakeCsrfProtector;
        $sut = $this->sut(["csrfProtector" => $csrfProtector, "db" => ["update" => 0]]);
        $request = new FakeRequest(["server" => ["REQUEST_TIME" => 1675205155]]);
        $sut($request, "do_edit");
        $this->assertTrue($csrfProtector->hasChecked());
    }

    public function testDoEditActionRedirectsOnSuccess()
    {
        $_POST = $this->dummyPost();
        $sut = $this->sut(["db" => ["update" => 1]]);
        $request = new FakeRequest(["server" => ["REQUEST_TIME" => 1675205155]]);
        $response = $sut($request, "do_edit");
        $this->assertEquals(
            "http://example.com/?realblog&admin=plugin_main&action=plugin_text",
            $response->location()
        );
    }

    public function testDoEditActionFailureIsReported(): void
    {
        $_POST = $this->dummyPost();
        $sut = $this->sut(["finder" => ["article" => $this->firstArticle()], "db" => ["update" => 0]]);
        $request = new FakeRequest(["server" => ["REQUEST_TIME" => 1675205155]]);
        $response = $sut($request, "do_edit");
        $this->assertEquals("Edit article", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testDoDeleteActionIsCsrfProtected()
    {
        $_POST = $this->dummyPost();
        $csrfProtector = new FakeCsrfProtector;
        $sut = $this->sut(["csrfProtector" => $csrfProtector, "db" => ["delete" => 0]]);
        $request = new FakeRequest(["server" => ["REQUEST_TIME" => 1675205155]]);
        $sut($request, "do_delete");
        $this->assertTrue($csrfProtector->hasChecked());
    }

   public function testDoDeleteActionRedirectsOnSuccess()
    {
        $_POST = $this->dummyPost();
        $sut = $this->sut(["db" => ["delete" => 1]]);
        $request = new FakeRequest(["server" => ["REQUEST_TIME" => 1675205155]]);
        $response = $sut($request, "do_delete");
        $this->assertEquals(
            "http://example.com/?realblog&admin=plugin_main&action=plugin_text",
            $response->location()
        );
    }

    public function testDoDeleteActionFailureIsReported(): void
    {
        $_POST = $this->dummyPost();
        $sut = $this->sut(["finder" => ["article" => $this->firstArticle()], "db" => ["delete" => 0]]);
        $request = new FakeRequest(["server" => ["REQUEST_TIME" => 1675205155]]);
        $response = $sut($request, "do_delete");
        $this->assertEquals("Delete article", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testDeleteSelectedActionRendersConfirmation()
    {
        $sut = $this->sut();
        $request = new FakeRequest(["get" => ["realblog_ids" => ["17", "4"]]]);
        $response = $sut($request, "delete_selected");
        Approvals::verifyHtml($response->output());
    }

    public function testDeleteSelectedActionReportsIfNothingIsSelected()
    {
        $sut = $this->sut();
        $request = new FakeRequest();
        $response = $sut($request, "delete_selected");
        Approvals::verifyHtml($response->output());
    }

    public function testChangeStatusActionRendersConfirmation()
    {
        $sut = $this->sut();
        $request = new FakeRequest(["get" => ["realblog_ids" => ["17", "4"]]]);
        $response = $sut($request, "change_status");
        Approvals::verifyHtml($response->output());
    }

    public function testChangeStatusActionReportsIfNothingIsSelected()
    {
        $sut = $this->sut();
        $request = new FakeRequest();
        $response = $sut($request, "change_status");
        Approvals::verifyHtml($response->output());
    }

    public function testDoDeleteSelectedActionIsCsrfProtected()
    {
        $_POST = ["realblog_ids" => ["17", "4"]];
        $csrfProtector = new FakeCsrfProtector;
        $sut = $this->sut(["csrfProtector" => $csrfProtector, "db" => ["bulkDelete" => 0]]);
        $request = new FakeRequest();
        $sut($request, "do_delete_selected");
        $this->assertTrue($csrfProtector->hasChecked());
    }

    public function testDoDeleteSelectedActionRedirectsOnSuccess()
    {
        $_POST = ["realblog_ids" => ["17", "4"]];
        $sut = $this->sut(["db" => ["bulkDelete" => 2]]);
        $request = new FakeRequest();
        $response = $sut($request, "do_delete_selected");
        $this->assertEquals(
            "http://example.com/?realblog&admin=plugin_main&action=plugin_text",
            $response->location()
        );
    }

    public function testDoDeleteSelectedActionReportsPartialSuccess()
    {
        $_POST = ["realblog_ids" => ["17", "4"]];
        $sut = $this->sut(["db" => ["bulkDelete" => 1]]);
        $request = new FakeRequest();
        $response = $sut($request, "do_delete_selected");
        $this->assertEquals("Delete selected articles", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testDoDeleteSelectedActionReportsFailure()
    {
        $_POST = ["realblog_ids" => ["17", "4"]];
        $sut = $this->sut(["db" => ["bulkDelete" => 0]]);
        $request = new FakeRequest();
        $response = $sut($request, "do_delete_selected");
        $this->assertEquals("Delete selected articles", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testDoChangeStatusActionIsCsrfProtected()
    {
        $_POST = ["realblog_ids" => ["17", "4"]];
        $csrfProtector = new FakeCsrfProtector;
        $sut = $this->sut(["csrfProtector" => $csrfProtector, "db" => ["bulkUpdate" => 0]]);
        $request = new FakeRequest();
        $sut($request, "do_change_status");
        $this->assertTrue($csrfProtector->hasChecked());
    }

    public function testDoChangeStatusActionRedirectsOnSuccess()
    {
        $_POST = ["realblog_ids" => ["17", "4"]];
        $sut = $this->sut(["db" => ["bulkUpdate" => 2]]);
        $request = new FakeRequest();
        $response = $sut($request, "do_change_status");
        $this->assertEquals("http://example.com/?realblog&admin=plugin_main&action=plugin_text", $response->location());
    }

    public function testDoChangeStatusActionReportsPartialSuccess()
    {
        $_POST = ["realblog_ids" => ["17", "4"]];
        $sut = $this->sut(["db" => ["bulkUpdate" => 1]]);
        $request = new FakeRequest();
        $response = $sut($request, "do_change_status");
        $this->assertEquals("Change article status", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testDoChangeStatusActionReportsFailure()
    {
        $_POST = ["realblog_ids" => ["17", "4"]];
        $sut = $this->sut(["db" => ["bulkUpdate" => 0]]);
        $request = new FakeRequest();
        $response = $sut($request, "do_change_status");
        $this->assertEquals("Change article status", $response->title());
        Approvals::verifyHtml($response->output());
    }

    private function sut($options = [])
    {
        return new MainAdminController(
            $this->conf($options["conf"] ?? []),
            $this->db($options["db"] ?? []),
            $this->finder($options["finder"] ?? []),
            $options["csrfProtector"] ?? new FakeCsrfProtector,
            $this->view(),
            $options["editor"] ?? new FakeEditor
        );
    }

    private function db($options = [])
    {
        $db = $this->createMock(DB::class);
        $db->expects(isset($options["bulkDelete"]) ? $this->once() : $this->never())
            ->method("deleteArticlesWithIds")->willReturn($options["bulkDelete"] ?? 0);
        $db->expects(isset($options["bulkUpdate"]) ? $this->once() : $this->never())
            ->method("updateStatusOfArticlesWithIds")->willReturn($options["bulkUpdate"] ?? 0);
        $db->expects(isset($options["insert"]) ? $this->once() : $this->never())
            ->method("insertArticle")->willReturn($options["insert"] ?? 0);
        $db->expects(isset($options["update"]) ? $this->once() : $this->never())
            ->method("updateArticle")->willReturn($options["update"] ?? 0);
        $db->expects(isset($options["delete"]) ? $this->once() : $this->never())
            ->method("deleteArticle")->willReturn($options["delete"] ?? 0);
        return $db;
    }

    private function finder($options = [])
    {
        $finder = $this->createStub(Finder::class);
        $finder->method("countArticlesWithStatus")->willReturn(count($options["articles"] ?? []));
        $finder->method('findArticlesWithStatus')->willReturn($options["articles"] ?? []);
        $finder->method('findById')->willReturn($options["article"] ?? null);
        $finder->method('findAllCategories')->willReturn(["cat1", "cat2"]);
        return $finder;
    }

    private function view()
    {
        return new View("./views/", XH_includeVar("./languages/en.php", 'plugin_tx')['realblog']);
    }

    private function conf($options = [])
    {
        $conf = XH_includeVar("./config/config.php", 'plugin_cf')['realblog'];
        return $options + $conf;
    }

    public function dummyPost(): array
    {
        return [
            'realblog_id' => "",
            'realblog_version' => "",
            'realblog_date' => "2023-02-01",
            'realblog_startdate' => "2023-02-01",
            'realblog_enddate' => "2024-02-01",
            'realblog_status' => "",
            'realblog_categories' => "",
            'realblog_title' => "",
            'realblog_headline' => "",
            'realblog_story' => "",
        ];
    }

    private function firstArticle(): FullArticle
    {
        return new FullArticle(
            1,
            1,
            1675205155,
            1675205155,
            0,
            1,
            "cat1",
            "Welcome!",
            "Welcome to my wonderful new blog",
            "Some lengthy blog post.",
            true,
            false
        );
    }

    private function articles()
    {
        return [new Article(
            1,
            strtotime("2023-01-31T22:45:55+00:00"),
            Article::PUBLISHED,
            "",
            "Welcome!",
            "Welcome to my wonderful new blog",
            true,
            true,
            false
        )];
    }
}
