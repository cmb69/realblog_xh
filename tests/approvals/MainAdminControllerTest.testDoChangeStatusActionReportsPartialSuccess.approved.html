<!-- realblog info message -->
<h1>Realblog â€“ Change article status</h1>
<p class="xh_warning">The status of 1 of 2 selected articles has been changed!</p><p><a href="/?realblog&amp;admin=plugin_main&amp;action=plugin_text">entries overview</a></p>
